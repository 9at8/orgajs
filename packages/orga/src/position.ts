export const map = (text: string) => {

  let cursor = 0
  const lines: number[] = [] // index of line starts
  do {
    lines.push(cursor)
    const nl = text.indexOf('\n', cursor)
    cursor = nl + 1
  } while (cursor > 0 && cursor < text.length)

  // const getLineRangeByIndex = (start: number) : Range => {
  //   const end = text.indexOf('\n', start)
  //   return {
  //     start,
  //     end: end >= 0 ? end + 1 : text.length,
  //   }
  // }

  const getLineRangeByLineNumber = (ln: number): Range | undefined => {
    if (ln + 1 >= lines.length) return undefined
    return {
      start: lines[ln],
      end: lines[ln + 1],
    }
  }

  const isLastLine = (ln: number) => {
    return ln >= lines.length
  }

  const toIndex = ({ line, column }: Point) : number => {
    return lines[line] + column
  }

  const substring = (position: Position) : string => {
    const start = toIndex(position.start)
    const end = toIndex(position.end)
    return text.substring(start, end)
  }

  const middle = (start, end) => {
    return start + Math.floor((end - start) / 2)
  }

  const findLine = (index: number, left: number, right: number) => {
    if (index < 0) return 0
    if (index >= text.length) return lines.length - 1
    const mid = middle(left, right)
    if (lines[mid] > index) return findLine(index, left, mid)
    if (lines[mid + 1] <= index) return findLine(index, mid, right)
    return mid
  }

  const location = (index: number): Point => {
    const line = findLine(index, 0, lines.length)
    const column = Math.min(index, text.length) - lines[line]
    return {
      line,
      column,
    }
  }

  const position = (): Position => {
    return {
      start: { line: 0, column: 0 },
      end: { line: lines.length - 1, column: text.length - lines[lines.length - 1] }
    }
  }

  const getLine = (ln: number, offset: number = 0) => {
    const range = getLineRangeByLineNumber(ln)
    if (!range) {
      console.log(`>>>>>> cant get line range`)
      return ''
    }
    return text.substring(range.start + offset, range.end)
  }

  const match = (
    pattern: RegExp,
    position: Position,
  ) : { position: Position, captures: string[] } | undefined => {
    const content = substring(position)
    if (!content) return undefined
    const match = pattern.exec(content)
    if (!match) return undefined
    const offset = toIndex(position.start)
    const captures = match.map(m => m)
    return {
      captures,
      position: {
        start: location(offset + match.index) as Point,
        end: location(offset + match.index + match[0].length) as Point,
      }
    }
  }

  const debug = () => {
    console.log({
      lines,
    })
  }

  return {
    getLineRangeByLineNumber,
    isLastLine,
    substring,
    getLine,
    location,
    position,
    match,
    toIndex,
    debug,
  }
}
