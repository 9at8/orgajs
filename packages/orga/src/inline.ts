import Node from './node'
import uri from './uri'
import { map } from './position'

const LINK_PATTERN = /\[\[([^\]]*)\](?:\[([^\]]*)\])?\]/m; // \1 => link, \2 => text
const FOOTNOTE_PATTERN = /\[fn:(\w+)\]/

const PRE = `[\\s\\({'"]|^`
const POST = `[\\s-\\.,:!?'\\)}]|$`
const BORDER = `[^,'"\\s]`

function markup(marker: string) {
  return RegExp(`(?<=(${PRE}))${marker}(${BORDER}(?:.*?(?:${BORDER}))??)${marker}(?=(${POST}.*))`, 'm')
}

export const parse = (text: string) => {
  text = _parse(LINK_PATTERN, text, (captures) => {
    return new Node(`link`)
      .with({ uri: uri(captures[0]), desc: captures[1] })
  })

  text = _parse(FOOTNOTE_PATTERN, text, (captures) => {
    return new Node(`footnote.reference`)
      .with({ label: captures[0] })
  })

  const markups = [
    { name: `bold`, marker: `\\*` },
    { name: `verbatim`, marker: `=` },
    { name: `italic`, marker: `/` },
    { name: `strikeThrough`, marker: `\\+` },
    { name: `underline`, marker: `_` },
    { name: `code`, marker: `~` },
  ]
  for (const { name, marker } of markups) {
    text = _parse(markup(marker), text, (captures) => {
      return new Node(name, captures[0])
    })
  }
  return text
}

function _parse(pattern, text, post) {
  if (typeof text === `string`) {
    const m = pattern.exec(text)
    if (!m) return [new Node(`text`).with({ value: text })]
    m.shift()
    const before = m.shift()
    const after = m.pop()
    let nodes: Node[] = []
    if ( before.length > 0 ) {
      nodes.push(new Node(`text`).with({ value: before }))
    }
    if (m.length > 0) {
      nodes.push(post(m))
      // nodes.push(new Node(name).with({ value: match }))
    }
    if (after) {
      nodes = nodes.concat(_parse(pattern, after, post))
    }
    return nodes
  }

  if (Array.isArray(text)) {
    return text.reduce((all, node) => {
      if (node.hasOwnProperty(`type`) && node.type !== `text`) {
        return all.concat(node)
      }
      return all.concat(_parse(pattern, node, post))
    }, [])
  }

  if (typeof text.value === `string`) {
    return _parse(pattern, text.value, post)
  }
  return undefined
}

const shift = (point: Point, offset: Point) => {
  return {
    line: point.line + offset.line,
    column: point.column + offset.column,
  }
}

const shiftRange = (range: Range, offset: number) => {
  return {
    start: range.start + offset,
    end: range.end + offset,
  }
}

const shiftPosition = (position: Position, offset: Point) => {
  return {
    start: shift(position.start, offset),
    end: shift(position.end, offset),
  }
}

const DEFAULT_TOKEN_NAME = 'text.plain'

export const tokenize = (text: string, offset: Point | undefined = undefined) : Token[] => {

  const { position, match, toIndex } = map(text)
  let tokens = [
    { name: DEFAULT_TOKEN_NAME, position: position() }
  ]

  const parse = (
    name: string,
    pattern: RegExp,
    content: Token[],
    data: (captures: string[]) => any = () => undefined,
  ): Token[] => {

    return content.reduce((all, token) => {
      if (token.name !== DEFAULT_TOKEN_NAME) return all.concat(token)
      const m = match(pattern, token.position)
      if (!m) return all.concat(token)
      if (toIndex(m.position.start) > toIndex(token.position.start)) {
        all.push({ name: DEFAULT_TOKEN_NAME, position: {
          start: token.position.start,
          end: m.position.start,
        } })
      }

      all.push({
        name,
        position: m.position,
        data: data(m.captures),
      })

      if (toIndex(m.position.end) < toIndex(token.position.end)) {
        const rest = parse(name, pattern, [
          {name: DEFAULT_TOKEN_NAME,
          position: {
            start: m.position.end,
            end: token.position.end,
          }}
        ])
        all = all.concat(rest)
      }

      return all
    }, [] as Token[])
  }

  tokens = parse('text.link', LINK_PATTERN, tokens, (captures) => ({
    uri: uri(captures[1]),
    description: captures[2],
  }))

  tokens = parse('text.footnote', FOOTNOTE_PATTERN, tokens, (captures) => ({
    label: captures[1],
  }))

  tokens = parse('text.bold', markup('\\*'), tokens)
  tokens = parse('text.verbatim', markup('='), tokens)
  tokens = parse('text.italic', markup('/'), tokens)
  tokens = parse('text.strikeThrough', markup('\\+'), tokens)
  tokens = parse('text.underline', markup('_'), tokens)
  tokens = parse('text.code', markup('~'), tokens)

  if (offset) {
    return tokens.map(token => ({
      name: token.name,
      position: shiftPosition(token.position, offset)
    }))
  }
  return tokens
}
