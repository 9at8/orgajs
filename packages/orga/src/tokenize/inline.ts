import Node from '../node'
import uri from '../uri'
import { map, isGreaterThan } from '../position'
import { Reader } from '../reader'


const LINK_PATTERN = /\[\[([^\]]*)\](?:\[([^\]]*)\])?\]/m; // \1 => link, \2 => text
const FOOTNOTE_PATTERN = /\[fn:(\w+)\]/

const PRE = `[\\s\\({'"]|^`
const POST = `[\\s-\\.,:!?'\\)}]|$`
const BORDER = `[^,'"\\s]`

function markup(marker: string) {
  return RegExp(`(?<=(${PRE}))${marker}(${BORDER}(?:.*?(?:${BORDER}))??)${marker}(?=(${POST}.*))`, 'm')
}

const shift = (point: Point, offset: Point) => {
  return {
    line: point.line + offset.line,
    column: point.column + offset.column,
  }
}

const shiftRange = (range: Range, offset: number) => {
  return {
    start: range.start + offset,
    end: range.end + offset,
  }
}

const shiftPosition = (position: Position, offset: Point) => {
  return {
    start: shift(position.start, offset),
    end: shift(position.end, offset),
  }
}

const DEFAULT_TOKEN_NAME = 'text.plain'

interface Props {
  reader: Reader;
  start?: Point;
  end?: Point;
}

export const tokenize = ({ reader, start, end } : Props) => {
  const { now, eol, match, substring } = reader
  const s = start || now()
  const e = end || eol()

  let tokens = [
    { name: DEFAULT_TOKEN_NAME, position: { start: s, end: e } }
  ]

  const parse = (
    name: string,
    pattern: RegExp,
    content: Token[],
    data: (captures: string[]) => any = () => undefined,
  ): Token[] => {

    return content.reduce((all, token) => {
      if (token.name !== DEFAULT_TOKEN_NAME) return all.concat(token)
      const m = match(pattern, token.position)
      if (!m) return all.concat(token)
      if (isGreaterThan(m.position.start, token.position.start)) {
        all.push({ name: DEFAULT_TOKEN_NAME, position: {
          start: token.position.start,
          end: m.position.start,
        } })
      }

      all.push({
        name,
        position: m.position,
        data: data(m.captures),
      })

      if (isGreaterThan(token.position.end, m.position.end)) {
        const rest = parse(name, pattern, [
          {name: DEFAULT_TOKEN_NAME,
          position: {
            start: m.position.end,
            end: token.position.end,
          }}
        ])
        all = all.concat(rest)
      }

      return all
    }, [] as Token[])
  }

  tokens = parse('text.link', LINK_PATTERN, tokens, (captures) => ({
    uri: uri(captures[1]),
    description: captures[2],
  }))

  tokens = parse('text.footnote', FOOTNOTE_PATTERN, tokens, (captures) => ({
    label: captures[1],
  }))

  tokens = parse('text.bold', markup('\\*'), tokens)
  tokens = parse('text.verbatim', markup('='), tokens)
  tokens = parse('text.italic', markup('/'), tokens)
  tokens = parse('text.strikeThrough', markup('\\+'), tokens)
  tokens = parse('text.underline', markup('_'), tokens)
  tokens = parse('text.code', markup('~'), tokens)
  return tokens
}
