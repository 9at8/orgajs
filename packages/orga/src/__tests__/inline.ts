import { parse, tokenize  } from '../inline'
import { map } from '../position'

const debug = (text: string) => {
  const { substring } = map(text)
  const tokens = tokenize(text)
  tokens.forEach(token => {
    console.log({
      ...token,
      content: substring(token.position) })
  })
}

describe('Inline Parsing', () => {
  it('recon single emphasis', () => {

    const str = 'hello *world*, welcome to *org-mode*. Have a nice *day*.'
    expect(tokenize(str)).toMatchSnapshot()
  })

  it('recon mixed emphasis', () => {

    const str = `
[[https://github.com/xiaoxinghu/orgajs][Here's]] to the *crazy* ones, the /misfits/, the _rebels_, the ~troublemakers~, the round pegs in the +round+ square holes...
`
    expect(tokenize(str)).toMatchSnapshot()
  })

  it('recon emphasises at different locations', () => {
    expect(tokenize(`one *two* three`)).toMatchSnapshot()
    expect(tokenize(`*one* two three`)).toMatchSnapshot()
    expect(tokenize(`one two *three*`)).toMatchSnapshot()
  })

  it('recon link', () => {
    expect(tokenize(`hello [[./image/logo.png]]`)).toMatchSnapshot()
    expect(tokenize(`hello [[Internal Link][link]]`)).toMatchSnapshot()
    expect(tokenize(`hello [[../image/logo.png][logo]]`)).toMatchSnapshot()
  })

  it('recon footnote reference', () => {
    expect(tokenize(`hello[fn:1] world.`)).toMatchSnapshot()
  })

  it('recon emphasises with 2 chars', () => {
    expect(tokenize(`*12*`)).toMatchSnapshot()
    expect(tokenize(`*1*`)).toMatchSnapshot()
  })

  it('recon invalid inline markups', () => {
    [ ',', '\'', '"', ' ' ].forEach(c => {
      expect(tokenize(`*${c}word*`)).toMatchSnapshot() // begin
      expect(tokenize(`*word${c}*`)).toMatchSnapshot() // end
    })
  })

  it('can handle something more complicated', () => {
    const content = `
Special characters =~= and =!=. Also =~/.this/path= and ~that~ thing.
`
    expect(tokenize(content)).toMatchSnapshot()
  })
})
