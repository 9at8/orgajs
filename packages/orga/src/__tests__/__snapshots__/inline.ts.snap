// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inline Parsing can handle something more complicated 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 19,
        "line": 1,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.verbatim",
    "position": Object {
      "end": Object {
        "column": 22,
        "line": 1,
      },
      "start": Object {
        "column": 19,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 27,
        "line": 1,
      },
      "start": Object {
        "column": 22,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.verbatim",
    "position": Object {
      "end": Object {
        "column": 30,
        "line": 1,
      },
      "start": Object {
        "column": 27,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 37,
        "line": 1,
      },
      "start": Object {
        "column": 30,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.verbatim",
    "position": Object {
      "end": Object {
        "column": 51,
        "line": 1,
      },
      "start": Object {
        "column": 37,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 56,
        "line": 1,
      },
      "start": Object {
        "column": 51,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.code",
    "position": Object {
      "end": Object {
        "column": 62,
        "line": 1,
      },
      "start": Object {
        "column": 56,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 70,
        "line": 1,
      },
      "start": Object {
        "column": 62,
        "line": 1,
      },
    },
  },
]
`;

exports[`Inline Parsing recon emphasises at different locations 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 4,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 9,
        "line": 0,
      },
      "start": Object {
        "column": 4,
        "line": 0,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 15,
        "line": 0,
      },
      "start": Object {
        "column": 9,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon emphasises at different locations 2`] = `
Array [
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 5,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 15,
        "line": 0,
      },
      "start": Object {
        "column": 5,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon emphasises at different locations 3`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 8,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 15,
        "line": 0,
      },
      "start": Object {
        "column": 8,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon emphasises with 2 chars 1`] = `
Array [
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 4,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon emphasises with 2 chars 2`] = `
Array [
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 3,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon footnote reference 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 5,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": Object {
      "label": "1",
    },
    "name": "text.footnote",
    "position": Object {
      "end": Object {
        "column": 11,
        "line": 0,
      },
      "start": Object {
        "column": 5,
        "line": 0,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 18,
        "line": 0,
      },
      "start": Object {
        "column": 11,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 2`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 3`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 4`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 5`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 6`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 7`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon invalid inline markups 8`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 7,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon link 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 6,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": Object {
      "description": undefined,
      "uri": Object {
        "location": "./image/logo.png",
        "protocol": "file",
        "raw": "./image/logo.png",
      },
    },
    "name": "text.link",
    "position": Object {
      "end": Object {
        "column": 26,
        "line": 0,
      },
      "start": Object {
        "column": 6,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon link 2`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 6,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": Object {
      "description": "link",
      "uri": Object {
        "location": "Internal Link",
        "protocol": "internal",
        "raw": "Internal Link",
      },
    },
    "name": "text.link",
    "position": Object {
      "end": Object {
        "column": 29,
        "line": 0,
      },
      "start": Object {
        "column": 6,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon link 3`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 6,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": Object {
      "description": "logo",
      "uri": Object {
        "location": "../image/logo.png",
        "protocol": "file",
        "raw": "../image/logo.png",
      },
    },
    "name": "text.link",
    "position": Object {
      "end": Object {
        "column": 33,
        "line": 0,
      },
      "start": Object {
        "column": 6,
        "line": 0,
      },
    },
  },
]
`;

exports[`Inline Parsing recon mixed emphasis 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 0,
        "line": 1,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": Object {
      "description": "Here's",
      "uri": Object {
        "location": "//github.com/xiaoxinghu/orgajs",
        "protocol": "https",
        "raw": "https://github.com/xiaoxinghu/orgajs",
      },
    },
    "name": "text.link",
    "position": Object {
      "end": Object {
        "column": 48,
        "line": 1,
      },
      "start": Object {
        "column": 0,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 56,
        "line": 1,
      },
      "start": Object {
        "column": 48,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 63,
        "line": 1,
      },
      "start": Object {
        "column": 56,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 74,
        "line": 1,
      },
      "start": Object {
        "column": 63,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.italic",
    "position": Object {
      "end": Object {
        "column": 83,
        "line": 1,
      },
      "start": Object {
        "column": 74,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 89,
        "line": 1,
      },
      "start": Object {
        "column": 83,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.underline",
    "position": Object {
      "end": Object {
        "column": 97,
        "line": 1,
      },
      "start": Object {
        "column": 89,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 103,
        "line": 1,
      },
      "start": Object {
        "column": 97,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.code",
    "position": Object {
      "end": Object {
        "column": 118,
        "line": 1,
      },
      "start": Object {
        "column": 103,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 142,
        "line": 1,
      },
      "start": Object {
        "column": 118,
        "line": 1,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.strikeThrough",
    "position": Object {
      "end": Object {
        "column": 149,
        "line": 1,
      },
      "start": Object {
        "column": 142,
        "line": 1,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 166,
        "line": 1,
      },
      "start": Object {
        "column": 149,
        "line": 1,
      },
    },
  },
]
`;

exports[`Inline Parsing recon single emphasis 1`] = `
Array [
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 6,
        "line": 0,
      },
      "start": Object {
        "column": 0,
        "line": 0,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 13,
        "line": 0,
      },
      "start": Object {
        "column": 6,
        "line": 0,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 26,
        "line": 0,
      },
      "start": Object {
        "column": 13,
        "line": 0,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 36,
        "line": 0,
      },
      "start": Object {
        "column": 26,
        "line": 0,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 50,
        "line": 0,
      },
      "start": Object {
        "column": 36,
        "line": 0,
      },
    },
  },
  Object {
    "data": undefined,
    "name": "text.bold",
    "position": Object {
      "end": Object {
        "column": 55,
        "line": 0,
      },
      "start": Object {
        "column": 50,
        "line": 0,
      },
    },
  },
  Object {
    "name": "text.plain",
    "position": Object {
      "end": Object {
        "column": 56,
        "line": 0,
      },
      "start": Object {
        "column": 55,
        "line": 0,
      },
    },
  },
]
`;
